### PR TITLE
Sharing map sharing statistics

### PR DESCRIPTION
**Pull request by danpoe**
This adds functionality to compute how many nodes are shared between a set of sharing maps.

### CODE REVIEW

File: unit/util/sharing_map.cpp [hunk 1]
```diff
@@ -6,11 +6,12 @@ Author: Daniel Poetzl
...
 \*******************************************************************/
 
-#define SHARING_MAP_INTERNAL_CHECKS
-#define SHARING_NODE_INTERNAL_CHECKS
+#define SM_INTERNAL_CHECKS
+#define SN_INTERNAL_CHECKS
 
 #include <climits>
 #include <random>
+#include <set>
```
Comment (tautschnig) [inline unit/util/sharing_map.cpp:14]: This seems unrelated to the topic of the PR?

File: src/util/sharing_node.h [hunk 2]
```diff
@@ -151,8 +153,6 @@ SN_TYPE_PAR_DEF class sharing_node_innert : public sharing_node_baset
 
   d_it &write_internal()
   {
-    SN_ASSERT(data.use_count() > 0);
-
```
Comment (tautschnig) [inline src/util/sharing_node.h:155]: Is there any test that shows that this is wrong (and is now correct)?

File: src/util/sharing_node.h [hunk 3]
```diff
@@ -257,8 +257,9 @@ SN_TYPE_PAR_DEF class sharing_node_innert : public sharing_node_baset
   leaft *place_leaf(const keyT &k, const valueT &v)
   {
     SN_ASSERT(is_container());
-
-    SN_ASSERT(as_const(this)->find_leaf(k) == nullptr);
+    // we need to check empty() first as the const version of find_leaf() must
+    // not be called on an empty node
+    SN_ASSERT(empty() || as_const(this)->find_leaf(k) == nullptr);
```
Comment (tautschnig) [inline src/util/sharing_node.h:262]: Again, is there a test?

File: src/util/sharing_map.h [hunk 4]
```diff
@@ -128,6 +134,7 @@ class sharing_mapt
   friend void sharing_map_copy_test();
   friend void sharing_map_collision_test();
   friend void sharing_map_view_test();
+  friend void sharing_map_sharing_stats_test();
```
Comment (tautschnig) [inline src/util/sharing_map.h]: At the risk of getting side-tracked: couldn't those tests be implemented by deriving a new child that provides more access? Those far-away friends are a bit of a pain point...

File: src/util/sharing_map.h [hunk 5]
```diff
@@ -317,6 +354,11 @@ class sharing_mapt
   void gather_all(const baset &n, const unsigned depth, delta_viewt &delta_view)
     const;
 
+  unsigned count_unmarked_nodes(
```
Comment (tautschnig) [inline src/util/sharing_map.h]: This should almost certainly be `std::size_t`

File: src/util/sharing_map.h [hunk 6]
```diff
@@ -317,6 +354,11 @@ class sharing_mapt
...
+    bool leafs_only,
+    std::set<void *> &marked,
```
Comment (tautschnig) [inline src/util/sharing_map.h]: Phew, what is this meant to be? What can be done using a set of `void` pointers?

**Review by tautschnig — COMMENTED** [review]
That's probably very good work, but greater clarity would be appreciated. See details, below.

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 7577e52c8e8944fc828f96890e98aa499371582d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/77894414

File: unit/util/sharing_map.cpp (hunk 1 above)
Comment (danpoe) [reply unit/util/sharing_map.cpp:14]: The newly introduced functions `count_unmarked_nodes()` and `get_sharing_stats()` use `std::set` which was not used before in the sharing map.

Ah, the comment might refer to the two changed defines. Yes, that's a fix that activates the internal assertions in the unit tests and is technically unrelated to the topic of this PR. Should I make a separate PR for this?

File: src/util/sharing_map.h (hunk 6 above)
Comment (danpoe) [reply src/util/sharing_map.h]: The set `marked` contains pointers to the nodes with `use_count >= 2` that have been visited already during the traversal. It's used to avoid exploring shared subtrees more than once.

The sharing map has two different types of nodes that don't have a common base class, so I used `void` pointers here to avoid having to keep two separate sets `marked` for both node types.

File: src/util/sharing_map.h (hunk 4 above)
Comment (danpoe) [reply src/util/sharing_map.h]: Do you mean deriving a new class from `sharing_mapt` in the unit tests which has public methods that wrap the protected members that we want to test? Alternatively, we could also derive a new class from `sharing_mapt` in the unit tests which then includes those friend declarations above.

What is the issue with the current approach though? Since the friends declared above are only from the unit tests it seems relatively safe.

File: src/util/sharing_node.h (hunk 2 above)
Comment (danpoe) [reply src/util/sharing_node.h:155]: This code was already exercised by several tests, but I used the wrong macro names to activate the sharing map/node internal assertions in the unit tests. Thus, none of the internal assertions were enabled and we didn't see the assertion failures. This is corrected by the first commit in this PR.

File: src/util/sharing_node.h (hunk 3 above)
Comment (danpoe) [reply src/util/sharing_node.h:262]: (see comment above)

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: ab4f984f100bf76a393e075b4730c1625fe6cd32).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/79729117

File: src/util/sharing_map.h (hunk 4 above)
Comment (tautschnig) [reply src/util/sharing_map.h]: Far-away friends are easy to lose touch with. In code terms, I effectively need to `git grep` to figure out who you are trying to be friends with. Maybe the solution is just adding a comment :-)  In practice, adding a new test currently requires touching this code, which really shouldn't be necessary. Hence my preference for deriving a new class from `sharing_mapt` in the unit tests, which would restrict any changes in unit tests to the unit test code.

File: src/util/sharing_map.h [hunk 7]
```diff
@@ -291,6 +298,36 @@ class sharing_mapt
...
+  /// Nodes that are part of n of the maps are counted n times.
+  ///
+  /// The num_unique_nodes field gives the number of unique nodes in the given
+  /// maps. A node that is part of several of the maps is only counted once.
+  ///
+  /// The num_leafs and num_unique_leafs fields are similar to the above but
+  /// only leafs are counted.
+  class sharing_map_statst
+  {
+  public:
```
Comment (tautschnig) [inline src/util/sharing_map.h]: I'd just use a `struct` here as all members are public.

**Review by tautschnig — APPROVED** [review]

File: src/util/sharing_map.h (hunk 4 above)
Comment (danpoe) [reply src/util/sharing_map.h]: Ah, I see. Yeah, I'll derive from `sharing_mapt` in the unit tests.

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 4ae8eb6756ad137f841708829c9458f5669349e8).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80509299
